ajax_with_jquery_exercise/script.js

```diff
@@ -0,0 +1,338 @@
+
+    $login.on('click', function(e){
+        // this just rolls out the form.
+
+        if (submitToggle === false){
+            submitToggle = true;
+            $slideScreen.slideDown();
+        } else {
+            submitToggle = false;
+            $slideScreen.slideUp();
+        }
+
+    })
+
+    // 2. Log in an existing user.
+
+    $loginForm.on('submit',function(e){
+
+        e.preventDefault();
+        // create new div. as a new last child to 
+        // send data to heroku app
+        // take token returned and stick into localstorage
+        $.ajax({
+            method: "POST",
+            url: "https://hn-favorites.herokuapp.com/login",
+            data: {email:$('#email').val(),password:$('#password').val()},
+            dataType: "json"
+        }).then(function(res){
+            localStorage.setItem("token", res.auth_token);
+            authToken = res.auth_token;
+            toggleNav();
+            favObj = JSON.parse(localStorage.getItem("favObj"));
+            loadTopStories(); // ----------- need to use local storage to retrieve
+        })
+        .catch(function(err){
+            alert("Error logging in. Please check your credentials and try again.")
+            $loginForm.trigger("reset")
+        });
+
+        $loginForm.trigger("reset");
+
+        submitToggle = false;
+        $slideScreen.slideUp();
+
+    });
+
+    // 3. Let a new user sign up.
+
+    $newUserForm.on('submit',function(e){
+
+        e.preventDefault();
+
+        $.ajax({
+            method: "POST",
+            url: "https://hn-favorites.herokuapp.com/signup",
+            data: {email:$('#newemail').val(),password:$('#newpassword').val()},
+            dataType: "json"
+        }).then(function(res){
+            localStorage.setItem("token", res.auth_token);
+            authToken = res.auth_token;
+            toggleNav();
+            favObj = {};
+        })
+        .catch(function(err){console.log(err)});
+
+        $newUserForm.trigger("reset");
+
+        submitToggle = false;
+        $slideScreen.slideUp();
+
+    });
+
+    // 4. Let logged in users log out.
+
+    $logout.on('click',function(e){
+        // remove auth from local storage.
+        localStorage.removeItem("token");
+        authToken = undefined;
+        toggleNav();
+        localStorage.setItem("favObj", JSON.stringify(favObj));
+
+        // return to top 20 stories page and empty all stars
+        favObj = {};
+        loadTopStories();
+        let $glyphs = $(".glyphicon");
+        $glyphs.removeClass("glyphicon-star");
+        $glyphs.removeClass("glyphicon-star-empty")
+        $glyphs.addClass("glyphicon-star-empty");
+    });
+
+    //  5. Let users add or remove favorites.
+   
+    $listSection.on('click',".glyphicon",function(e){
+        if ($(e.target).hasClass("glyphicon-star-empty")){
+            // making a link a favorite
+
+            let byVar = $(e.target).parent().children().eq(3).children().eq(0).attr('data-by');
```

Comment:
save `$(e.target).parent().children()` in a variable to cache it. It's expensive to keep going to the DOM and doing this traversal/searching